File: basic_python/practices/Data Structures/01. Array & Hashing/Top K Frequent Elements/top_K_freq_elm.py
class topKFrequent:
    def hashmap_sorting(self, nums: list[int], k: int) -> list[int]:
        hashmap = {}
        for num in nums:
            cnt = 0
            if num in hashmap:
                cnt = cnt + 1
            hashmap[num] = hashmap.get(num, 0) + 1
        res = sorted(hashmap,
                    key = hashmap.get,
                    reverse = True
                    )[:k]
        return res
    
    # Approach 2:
    def min_heap(self, nums: list[int], k: int) -> list[int]:
        import heapq

        count = {}
        for num in nums:
            count[num] = 1 + count.get(num, 0)

        heap = []
        for num in count.keys():
            heapq.heappush(heap, (count[num], num))
            if len(heap) > k:
                heapq.heappop(heap)

        res = []
        for i in range(k):
            res.append(heapq.heappop(heap)[1])
        return res

File: basic_python/practices/Data Structures/01. Array & Hashing/Top K Frequent Elements/top_K_freq_elm.cpp
#include <vector>
#include <iostream>
#include <unordered_map>
#include <algorithm>
#include <queue>
using namespace std;

class TopKFrequent {
public:

    // ---------------------------------------------------------
    // Approach 1: Sort by frequency (O(n log n))
    // ---------------------------------------------------------
    vector<int> using_sorting(vector<int>& nums, int k) {
        unordered_map<int, int> count;
        for (int num : nums) {
            count[num]++;
        }

        vector<pair<int, int>> arr; // (freq, num)
        for (const auto& p : count) {
            arr.push_back({p.second, p.first});
        }

        sort(arr.rbegin(), arr.rend());  // sort descending on freq

        vector<int> res;
        for (int i = 0; i < k; ++i) {
            res.push_back(arr[i].second);
        }
        return res;
    }

    // ---------------------------------------------------------
    // Approach 2: Min Heap (O(n log k))
    // ---------------------------------------------------------
    vector<int> using_minheap(vector<int>& nums, int k) {
        unordered_map<int, int> count;
        for (int num : nums) {
            count[num]++;
        }

        // min-heap storing pair (frequency, number)
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> heap;

        for (auto& entry : count) {
            heap.push({entry.second, entry.first});
            if (heap.size() > k) {
                heap.pop();
            }
        }

        vector<int> res;
        while (!heap.empty()) {
            res.push_back(heap.top().second);
            heap.pop();
        }

        return res;
    }
};

int main() {
    TopKFrequent sol;

    int n, k;
    std::cout << "Enter number of elements: ";
    std::cin >> n;
    
    std::vector<int> nums(n);
    std::cout << "Enter " << n << " elements separated by spaces: ";
    for (int i = 0; i < n; i++) {
        std::cin >> nums[i];
    }

    std::cout << "Enter k: ";
    std::cin >> k;

    std::vector<int> result = sol.using_sorting(nums, k);
    // Or use using_minheap(nums, k) if you prefer

    std::cout << "Top " << k << " frequent elements: ";
    for (int num : result) {
        std::cout << num << " ";
    }
    std::cout << "\n";

    return 0;
}